design/weak-imports.md

```diff
@@ -0,0 +1,60 @@
+Weak Imports
+============
+
+It can be useful for WASI programs to weakly depend on certain functions in a
+WASI API.  For example, if a WASI API evolves to include a new function a
+program might want to continue to run on older systems that don't yet support
+the new function.  In this case a weak import mechanism allows the program to
+run on older systems and detect the presence of the function at runtime.
+
+WebAssembly itself does not currently provide a mechanism for weak imports.
+There is some discussion of adding it the spec [^1], and WASI would
+likely use such a feature if/when it is added.
+
+This document describes the mechanism used by WASI to achieve a form of weak
+import on top of the existing primitives.  Currently this is only defined for
+function imports.
```

Comment:
We should also mention that this feature is not related to weak GC references, since we've already had one question about this.